### PR TITLE
[15 min fix] Fix some autoloading deprecation warnings

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

When a console is load from scratch (`spring stop; bundle exec rails console`), Rails outputs quite a long deprecation message having to do with [autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html) and the new autloader that was added in Rails 6.

Addressing those deprecations, with the help of the [official guide on autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html) and [discourse's middlewares initializer](https://github.com/discourse/discourse/blob/dc6b547ed89f652b5406489d76140b76cf8e0d1d/config/initializers/200-first_middlewares.rb) removed some deprecations.

Unfortunately, I couldn't figure out how to deal with `ForemStatsClient` which is an odd pre-initialized constant we have lying around. 

I tried to use [Rails.application.reloader.to_prepare](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#autoloading-when-the-application-boots) to have the autoloader "digest" it correctly like we do in other cases *but*

https://github.com/forem/forem/blob/ff244a31eff9d8ad270dd7b0aa4c6ed5eac03223/app/lib/forem_stats_drivers/datadog_driver.rb#L17

has to access `config.middleware` which is already frozen when `.to_prepare` is called. The other thing I tried was to move the initialization inside `app/lib` but then it's not eager loaded correctly in production mode.

I wonder if there's something to be done to improve the code and make Rails happy 😕 

## QA Instructions, Screenshots, Recordings

You can run:

```
RAILS_ENV=production bundle exec rails middleware
```

to make sure `Middlewares::SetCookieDomain` is in the list.

You can run:

```
RAILS_ENV=test bundle exec rails middleware
```

to make sure `Middlewares::SetTimeZone` is in the list.

An additional check is to load the console in production mode and make sure there are no ddtrace related warnings: `RAILS_ENV=production bundle exec rails console`

Regarding deprecations, we went from:

```shell
Unset ENV variable: IMGPROXY_KEY.
Unset ENV variable: IMGPROXY_SALT.
W, [2021-05-18T18:08:12.386802 #47598]  WARN -- ddtrace: [ddtrace] Unable to patch Datadog::Contrib::RestClient::Integration (Available?: true, Loaded? false, Compatible? true, Patchable? false)
DEPRECATION WARNING: Initialization autoloaded the constants RateLimitChecker, ForemStatsDriver, ForemStatsDrivers, ActsAsForemStatsDriver, and ForemStatsDrivers::DatadogDriver.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload RateLimitChecker, for example,
the expected changes won't be reflected in that stale Class object.

These autoloaded constants have been unloaded.

In order to autoload safely at boot time, please wrap your code in a reloader
callback this way:

    Rails.application.reloader.to_prepare do
      # Autoload classes and modules needed at boot time here.
    end

That block runs when the application boots, and every time there is a reload.
For historical reasons, it may run twice, so it has to be idempotent.

Check the "Autoloading and Reloading Constants" guide to learn more about how
Rails autoloads and reloads.
 (called from <main> at /Users/rhymes/Development/forem/config/environment.rb:5)
```

to 

```shell
Unset ENV variable: IMGPROXY_KEY.
Unset ENV variable: IMGPROXY_SALT.
DEPRECATION WARNING: Initialization autoloaded the constants ForemStatsDriver, ForemStatsDrivers, ActsAsForemStatsDriver, and ForemStatsDrivers::DatadogDriver.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload ForemStatsDriver, for example,
the expected changes won't be reflected in that stale Class object.

These autoloaded constants have been unloaded.

In order to autoload safely at boot time, please wrap your code in a reloader
callback this way:

    Rails.application.reloader.to_prepare do
      # Autoload classes and modules needed at boot time here.
    end

That block runs when the application boots, and every time there is a reload.
For historical reasons, it may run twice, so it has to be idempotent.

Check the "Autoloading and Reloading Constants" guide to learn more about how
Rails autoloads and reloads.
 (called from <main> at /Users/rhymes/Development/forem/config/environment.rb:5)
```

at startup.

You can still load the app, click around and make sure everything is working.

## Added tests?

- [ ] Yes
- [x] No, and this is why: code has changed and tests need to pass as they are
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

File: app/lib/forem_stats_drivers/datadog_driver.rb [hunk 1]
```diff
@@ -1,28 +1,32 @@
 require "httpclient"
+require "rest-client"
```
Comment (rhymes) [inline app/lib/forem_stats_drivers/datadog_driver.rb:2]: `rest-client` is not a transitive dependency, so it needs to be required explicitly for the `c.use :rest_client` instruction to work

File: app/lib/forem_stats_drivers/datadog_driver.rb [hunk 2]
```diff
@@ -1,28 +1,32 @@
...
         c.use :httpclient, split_by_domain: false
         c.use :httprb, split_by_domain: true
+        c.use :rails
+        c.use :redis, service_name: "redis", describes: { url: ENV["REDIS_URL"] }
+        c.use :redis, service_name: "redis-rpush", describes: { url: ENV["REDIS_RPUSH_URL"] }
+        c.use :redis, service_name: "redis-sessions", describes: { url: ENV["REDIS_SESSIONS_URL"] }
+        c.use :redis, service_name: "redis-sidekiq", describes: { url: ENV["REDIS_SIDEKIQ_URL"] }
         c.use :rest_client
-        c.use :concurrent_ruby
+        c.use :sidekiq
```
Comment (rhymes) [inline app/lib/forem_stats_drivers/datadog_driver.rb:27]: just sorted these

File: app/lib/middlewares/set_cookie_domain.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+module Middlewares
```
Comment (rhymes) [inline app/lib/middlewares/set_cookie_domain.rb:1]: added a `Middlewares` module for grouping

File: config/application.rb [hunk 4]
```diff
@@ -63,8 +63,6 @@ class Application < Rails::Application
     # Therefore we disable "per_form_csrf_tokens" for the time being.
     config.action_controller.per_form_csrf_tokens = false
 
-    config.middleware.use SetCookieDomain
```
Comment (rhymes) [inline config/application.rb:66]: despite being here in the current code, `SetCookieDomain` is only actually used in `production`:

https://github.com/forem/forem/blob/ff244a31eff9d8ad270dd7b0aa4c6ed5eac03223/app/middlewares/set_cookie_domain.rb#L9-L14

As you can see, in any other environment, it's just a pass through call


no need to insert a global middleware when we don't use it, see https://github.com/forem/forem/pull/13793/files#diff-4493b648c41705909d0895e7af5b09ee416e774adf53b042e837975cf08d3a05

File: config/environments/test.rb [hunk 5]
```diff
@@ -11,10 +11,6 @@
 
   config.cache_classes = true
 
-  # Include middleware to ensure timezone for browser requests for Capybara specs
-  # matches the random zonebie timezone set at the beginning of our spec run
-  config.middleware.use SetTimeZone
```
Comment (rhymes) [inline config/environments/test.rb:16]: moved to `config/initializers/middlewares.rb`

File: config/initializers/honeybadger.rb [hunk 6]
```diff
@@ -1,25 +1,26 @@
...
-  ActiveRecord::RecordNotFound,
-  Pundit::NotAuthorizedError,
-  RateLimitChecker::LimitReached,
-].freeze
+  honeybadger_exceptions_to_ignore = [
+    ActiveRecord::QueryCanceled,
+    ActiveRecord::RecordNotFound,
+    Pundit::NotAuthorizedError,
+    RateLimitChecker::LimitReached,
+  ]
```
Comment (rhymes) [inline config/initializers/honeybadger.rb:22]: these shouldn't be constants, they're only used here and they actively create problems when in development Rails tries to reload these, finding them already defined

**Comment by citizen428** [comment]
I don't see any lengthy deprecation warnings when I run `spring stop; bundle exec rails console`:
  
image

Somehow this reminds me of the PR I did for `rails data_updates:status` where you got an `ActiveRecord` log message I never saw. Now I wonder again if you somehow have some extra verbose logging enabled or I have something disabled. ❓

**Comment by rhymes** [comment]
@citizen428 ahhaha I don't know. I see in your screenshot that you do have the `Unable to patch...` ddtrace warning though

**Comment by djuber** [comment]
Just to second @citizen428's point - I get the same result on main

spring stop; rails console

**Comment by maestromac** [comment]
Same-ish here. I don't use spring
```
╰─ bundle exec rails console
W, [2021-05-19T15:38:23.750251 #2931797]  WARN -- ddtrace: [ddtrace] Unable to patch Datadog::Contrib::RestClient::Integration (Available?: true, Loaded? false, Compatible? true, Patchable? false)
/home/mac/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/zeitwerk-2.4.2/lib/zeitwerk/loader.rb:590: warning: constant ::Data is deprecated
Running via Spring preloader in process 2932002
Loading development environment (Rails 6.1.3.2)
[1] pry(main)>
```

**Comment by citizen428** [comment]
@maestroc 

> I don't use spring

I think Rails disagrees 😉 

2021-05-20 09_20_42- 15 min fix  Fix some autoloading deprecation warnings by rhymes · Pull Request

**Comment by maestromac** [comment]
> @maestroc
> 
> > I don't use spring
> 
> I think Rails disagrees wink
> 
> 2021-05-20 09_20_42- 15 min fix Fix some autoloading deprecation warnings by rhymes · Pull Request

:rofl: oh how embarrassing. I typically use the non-spring binstub command, `bin/rails s`. I happen to run bundle exec this time to follow Rhyme's QA steps.

**Review by djuber — APPROVED** [review]

**Pull request merged**